include/unifex/detail/debug_async_scope.hpp

```diff
@@ -0,0 +1,209 @@
+      }
+      UNIFEX_CATCH(...) {
+        unifex::set_error(std::move(receiver), std::current_exception());
+      }
+    });
+  }
+
+  template <typename E>
+  void set_error(E&& e) noexcept {
+    op_->complete([&e](Receiver&& receiver) noexcept {
+      unifex::set_error(std::move(receiver), static_cast<E&&>(e));
+    });
+  }
+
+  void set_done() noexcept { op_->complete(unifex::set_done); }
+};
+
+template <typename Sender, typename Receiver>
+struct _operation_impl final {
+  struct type;
+};
+
+template <typename Sender, typename Receiver>
+using debug_op = typename _operation_impl<Sender, Receiver>::type;
+
+template <typename Sender, typename Receiver>
+struct _operation_impl<Sender, Receiver>::type final
+  : _operation<Receiver>::type {
+  using base_op_t = typename _operation<Receiver>::type;
+  using receiver_t = typename _receiver<Receiver>::type;
+
+  template <typename Sender2, typename Receiver2>
+  type(
+      debug_op_list* ops,
+      Sender2&& sender,
+      Receiver2&& receiver) noexcept  // TODO
+    : base_op_t(typeid(type), ops, static_cast<Receiver2&&>(receiver))
+    , op_(unifex::connect(static_cast<Sender2&&>(sender), receiver_t{this})) {}
+
+  type(type&&) = delete;
+
+  friend void tag_invoke(tag_t<start>, type& op) noexcept {
+    op.ops_->register_debug_operation(&op);
+    unifex::start(op.op_);
+  }
+
+private:
+  using op_t = connect_result_t<Sender, receiver_t>;
+  op_t op_;
+};
+
+template <typename Sender>
+struct _debug_scope_sender final {
+  struct type;
+};
+
+template <typename Sender>
+struct _debug_scope_sender<Sender>::type final {
+  template <
+      template <typename...>
+      typename Variant,
+      template <typename...>
+      typename Tuple>
+  using value_types = sender_value_types_t<Sender, Variant, Tuple>;
+
+  template <template <typename...> typename Variant>
+  using error_types =
+      sender_error_types_t<Sender, Variant>;  // TODO exception_ptr
+
+  static constexpr bool sends_done = sender_traits<Sender>::sends_done;
+
+  template <typename Sender2>
+  type(Sender2&& sender, debug_op_list* ops) noexcept(
+      std::is_nothrow_constructible_v<Sender, Sender2>)
+    : ops_(ops)
+    , sender_(static_cast<Sender2&&>(sender)) {}
+
+  template(typename Self, typename Receiver)          //
+      (requires same_as<type, remove_cvref_t<Self>>)  //
+      friend auto tag_invoke(
+          tag_t<connect>, Self&& self, Receiver&& receiver) noexcept {  // TODO
+    return debug_op<Sender, remove_cvref_t<Receiver>>{
+        static_cast<Self&&>(self).ops_,
+        static_cast<Self&&>(self).sender_,
+        static_cast<Receiver&&>(receiver)};
+  }
+
+private:
+  debug_op_list* ops_;
+  UNIFEX_NO_UNIQUE_ADDRESS Sender sender_;
+};
+
+}  // namespace _debug_async_scope
+
+using _debug_async_scope::debug_op_list;
+template <typename Sender>
+using debug_scope_sender =
+    typename _debug_async_scope::_debug_scope_sender<Sender>::type;
+
+}  // namespace unifex::detail
```

Comment:
We should probably merge as-is, but this _Sender_ wrapper is kinda-sorta general purpose so it might make sense to elevate it out of the `detail` namespace.